batch/sql/add-user-resources-ready-cores-token.sql

```diff
@@ -0,0 +1,176 @@
+    tmp_resources.n_ready_jobs,
+    tmp_resources.ready_cores_mcpu
+  FROM tmp_resources
+  WHERE NOT closed;
+
+DELIMITER $$
+
+DROP TRIGGER IF EXISTS jobs_after_update;
+CREATE TRIGGER jobs_after_update AFTER UPDATE ON jobs
+FOR EACH ROW
+BEGIN
+  DECLARE in_user VARCHAR(100);
+  DECLARE rand_token INT;
+
+  SELECT user INTO in_user from batches
+  WHERE id = NEW.batch_id;
+
+  SET rand_token = FLOOR(RAND() * @n_tokens);
+
+  IF OLD.state = 'Ready' THEN
+    INSERT INTO user_resources (user, token) VALUES (in_user, rand_token)
+    ON DUPLICATE KEY UPDATE
+      n_ready_jobs = n_ready_jobs - 1,
+      ready_cores_mcpu = ready_cores_mcpu - OLD.cores_mcpu;
+
+    INSERT INTO ready_cores (token, ready_cores_mcpu) VALUES (rand_token, 0)
+    ON DUPLICATE KEY UPDATE
+      ready_cores_mcpu = ready_cores_mcpu - OLD.cores_mcpu;
+  END IF;
+
+  IF NEW.state = 'Ready' THEN
+    INSERT INTO user_resources (user, token) VALUES (in_user, rand_token)
+    ON DUPLICATE KEY UPDATE
+      n_ready_jobs = n_ready_jobs + 1,
+      ready_cores_mcpu = ready_cores_mcpu + NEW.cores_mcpu;
+
+    INSERT INTO ready_cores (token, ready_cores_mcpu) VALUES (rand_token, 0)
+    ON DUPLICATE KEY UPDATE
+      ready_cores_mcpu = ready_cores_mcpu + NEW.cores_mcpu;
+  END IF;
+
+  IF OLD.state = 'Running' THEN
+    INSERT INTO user_resources (user, token) VALUES (in_user, rand_token)
+    ON DUPLICATE KEY UPDATE
+      n_running_jobs = n_running_jobs - 1,
+      running_cores_mcpu = running_cores_mcpu - OLD.cores_mcpu;
+  END IF;
+
+  IF NEW.state = 'Running' THEN
+    INSERT INTO user_resources (user, token) VALUES (in_user, rand_token)
+    ON DUPLICATE KEY UPDATE
+      n_running_jobs = n_running_jobs + 1,
+      running_cores_mcpu = running_cores_mcpu + NEW.cores_mcpu;
+  END IF;
+END $$
+
+DROP PROCEDURE IF EXISTS close_batch;
+CREATE PROCEDURE close_batch(
+  IN in_batch_id BIGINT,
+  IN in_timestamp BIGINT
+)
+BEGIN
+  DECLARE cur_batch_closed BOOLEAN;
+  DECLARE expected_n_jobs INT;
+  DECLARE staging_n_jobs INT;
+  DECLARE staging_n_ready_jobs INT;
+  DECLARE staging_ready_cores_mcpu INT;
+  DECLARE cur_user VARCHAR(100);
+
+  START TRANSACTION;
+
+  SELECT n_jobs, closed INTO expected_n_jobs, cur_batch_closed FROM batches
+  WHERE id = in_batch_id AND NOT deleted;
+
+  IF cur_batch_closed = 1 THEN
+    COMMIT;
+    SELECT 0 as rc;
+  ELSEIF cur_batch_closed = 0 THEN
+    SELECT SUM(n_jobs), SUM(n_ready_jobs), SUM(ready_cores_mcpu)
+    INTO staging_n_jobs, staging_n_ready_jobs, staging_ready_cores_mcpu
+    FROM batches_staging
+    WHERE batch_id = in_batch_id
+    FOR UPDATE;
+
+    SELECT user INTO cur_user FROM batches WHERE id = in_batch_id;
+
+    IF staging_n_jobs = expected_n_jobs THEN
+      UPDATE batches SET closed = 1 WHERE id = in_batch_id;
+      UPDATE batches SET time_completed = in_timestamp
+        WHERE id = in_batch_id AND n_completed = batches.n_jobs;
+
+      INSERT INTO ready_cores (token, ready_cores_mcpu) VALUES (0, 0)
```

Comment:
Unnecessary if you COALESCE the SUMs.  I think we should always COALESCE the SUMs.  I think the SQL empty SUM behavior was the wrong call.